### PR TITLE
Add the concept of rigid typevars for enforcing polymorpic signatures

### LINKED ISSUE #301: Invalidate type restriction in definition blocks w.r.t their signature
Currently when defining a definition block with a type variable, the result type can be restricted to a concrete type (or have additional type classes).
However, a result type should be *at least as generic* as the definition's signature.

A possible fix would be to check if the type feeding into the result anchor does not add restrictions to the signature's result type when unified.

Example of an offending type restriction:
shot_1226

### PR DESCRIPTION
**Pull request by comonoidial**
This resolves #301

### CODE REVIEW

File: Code/src/main/java/nl/utwente/viskell/haskell/type/TypeVar.java [hunk 1]
```diff
@@ -79,23 +98,42 @@ private ConcreteType get() {
...
 
         /**
          * Deeply unifying all aspects of type instances
          * @param the other type instance.
-         * @throws HaskellTypeError if the combined constraint set of the typeva instances is not satisfiable.
+         * @throws HaskellTypeError if the combined constraint set of the typevar instances is not satisfiable.
+         * @throws HaskellTypeError if a rigid type variable gets unified with more constraints or another rigid type variable
          */
         private void unifyWith(TypeInstance other) throws HaskellTypeError {
+            if (this == other) {
```
Comment (Flumble) [inline Code/src/main/java/nl/utwente/viskell/haskell/type/TypeVar.java]: Does it have to be the same reference or an isomorphic type?

Comment (comonoidial) [inline Code/src/main/java/nl/utwente/viskell/haskell/type/TypeVar.java]: Same reference as that is the exception to the condition 4 lines further.

Comment (Flumble) [inline Code/src/main/java/nl/utwente/viskell/haskell/type/TypeVar.java]: If both are rigid and isomorphic, they shouldn't produce an exception, right? A rigid ::a and ::b can live happily together.

Comment (comonoidial) [inline Code/src/main/java/nl/utwente/viskell/haskell/type/TypeVar.java]: I don't think so, the type a -> a is not the same as the type a -> b

**Pull request merged**